src/core_modules/capture-core/components/WidgetTrackedEntityRelationship/hooks.js

```diff
@@ -0,0 +1,66 @@
+// @flow
+import { useMemo, useRef, useState } from 'react';
+import {
+    getCachedResourceAsync,
+} from '../../MetaDataStoreUtils/MetaDataStoreUtils';
+import { userStores } from '../../storageControllers/stores';
+import type { RelationshipType } from './WidgetTrackedEntityRelationship.types';
+
+export const useRelationshipTypes = () => {
+    const [cachedTypes, setCachedTypes] = useState();
+    const [loading, setLoading] = useState(true);
+    const error = useRef();
+    useMemo(() => {
+        getCachedResourceAsync(userStores.RELATIONSHIP_TYPES)
```

Comment:
Hey, 
Why do you use the data from the Indexed DB here? To make the widget self-contained you can directly call the API using a hook. Something like this: 
```
const { error, loading, data } = useDataQuery(
    useMemo(
        () => ({
            relationshipTypes: {
                resource: 'relationshipTypes',
                params: () => ({
                    fields: 'id,displayName,bidirectional,fromToName,toFromName,fromConstraint[*],toConstraint[*],access[*]',
                }),
            },
        }),
        [],
    ),
);

```
Thank you!

Reply:
I see that I used the term self-contained in the JIRA-issue; not really the term I was looking for, so I've updated the JIRA description. The idea is to pass in the relationship types to the Widget (See more info in this ticket: https://jira.dhis2.org/browse/DHIS2-12362 that @jasminenguyennn is working on). The relationship types passed in should also be used in the "add" part that @eirikhaugstulen is working on here. I.e, we shouldn't get the relationshipTypes in the Widget itself (So at some point we need to coordinate)

Concerning metadata retrieval in general: For the time being I think it makes sense to grab it from our own indexedDB cache, because:
1) We have an established pattern that a) ensures we have updated cache values b) that won't make unnecessary requests to the api.
2) Currently, we will often need to use refetch() from useDataQuery to get the data. refetch() will skip the app-runtime cache and redo the api request. In this particular example we won't have that problem, but often we will need to get some new data if one of the props changes (programId, teiId etc.).

When the app-runtime adds offline-caching we can experiment with it to see if it meets our requirements. If/when it meets our requirements when can lay out a plan to change our approach. I would really like not to be alpha testers for this since we don't need the extra work and I think it would be beneficial for us to see how this works in another app first (I have some concerns; especially how will they ensure the cache is up to date)

We should probably have a conversion about this, so it is clear to the whole team. I think it would be great if @Bekkalizer also joined. I will set something up when I get back.